### PR TITLE
Behavior change of history passive task processing

### PR DESCRIPTION
**Pull request by wxing1292**
* History task processing will verify whether to process task only once
* WorkflowTimeout / DeleteHistory / ArchiveHistory / CloseExecution passive tasks are guaranteed to be executed.

### CODE REVIEW

File: service/history/replicatorQueueProcessor.go [hunk 1]
```diff
@@ -105,11 +111,17 @@ func newReplicatorQueueProcessor(shard ShardContext, historyCache *historyCache,
 	return processor
 }
 
-func (p *replicatorQueueProcessorImpl) process(qTask queueTaskInfo) (int, error) {
+func (p *replicatorQueueProcessorImpl) getTaskFilter() queueTaskFilter {
+	return p.replicationTaskFilter
+}
+
+func (p *replicatorQueueProcessorImpl) process(qTask queueTaskInfo, shouldProcessTask bool) (int, error) {
```
Comment (andrewjdawson2016) [inline service/history/replicatorQueueProcessor.go]: To make it clear that shouldProcessTask is not used can we write:

`func (p *replicatorQueueProcessorImpl) process(qTask queueTaskInfo, _ bool) (int, error) {`

Comment (vancexu) [inline service/history/replicatorQueueProcessor.go]: I am confused why `shouldProcessTask` passed in here. if `! shouldProcessTask`, shouldn't we just not call process?

File: service/history/queueProcessor.go [hunk 2]
```diff
@@ -271,12 +271,42 @@ func (p *queueProcessorBase) retryTasks() {
...
+	}
+
+FilterLoop:
+	for {
+		select {
+		case <-p.shutdownCh:
+			// this must return without ack
+			return
+		default:
+			shouldProcessTask, err = p.processor.getTaskFilter()(task)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/queueProcessor.go]: the task filter will be used by both active and passive logic, so isTaskInStandbyDC is not a good name
the getTaskFilter function returns a function, so `shouldProcessTask(task)` will be something like `shouldProcessTask()(task)`, which is weird?

File: service/history/replicatorQueueProcessor.go (hunk 1 above)
Comment (wxing1292) [reply service/history/replicatorQueueProcessor.go]: `WorkflowTimeout / DeleteHistory / ArchiveHistory / CloseExecution passive tasks are guaranteed to be executed.` meaning even if the task is consider active, passive logic will still process that task

File: service/history/queueProcessor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/queueProcessor.go]: active side will provide an different task filter compared to passive processing logic

[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/queueProcessor.go]: plus failover logic has its own task filter

**Pull request merged**